framework/include/utils/RankTwoTensor.h

```diff
@@ -262,149 +550,703 @@ class RankTwoTensorTempl : public TensorValue<T>
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   *
+   * When `input.size() == 3`, the vector values are used to fill the diagonal components of the
+   * second order tensor:
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+   * RankTwoTensor A;
+   * A.fillFromInputVector({1, 2, 3});
+   * // A = [ 1 0 0
+   * //       0 2 0
+   * //       0 0 3 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   *
+   * When `input.size() == 6`, the second order tensor is filled symmetrically using the Voigt
+   * notation:
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+   * RankTwoTensor A;
+   * A.fillFromInputVector({1, 2, 3, 4, 5, 6});
+   * // A = [ 1 6 5
+   * //       6 2 4
+   * //       5 4 3 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   *
+   * When `input.size() == 9`, all components of the second order tensor are filled in a
+   * column-major fashion:
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+   * RankTwoTensor A;
+   * A.fillFromInputVector({1, 2, 3, 4, 5, 6, 7, 8, 9});
+   * // A = [ 1 4 7
+   * //       2 5 8
+   * //       3 6 9 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
+  void fillFromInputVector(const std::vector<T> & input, FillMethod fill_method = autodetect);
+
+  /**
+   * @brief The smart mutator that determines how to fill the second order tensor based on the order
+   * of the scalar_variable.
+   *
+   * @param scalar_variable The input scalar variable. Supported orders are FIRST, THIRD, and SIXTH.
+   *
+   * When `scalar_variable.size() == 1`, the scalar value is used to fill the very first component
+   * of the second order tensor:
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+   & // Suppose v[0] = 1
+   * RankTwoTensor A;
+   * A.fillFromScalarVariable(v);
+   * // A = [ 1 0 0
+   * //       0 0 0
+   * //       0 0 0 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   *
+   * When `input.size() == 3`, the scalar values are used to fill the in-plane components of the
+   second order tensor using the Voigt notation:
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+   * // Suppose v[0] = 1
+   * //         v[1] = 2
+   * //         v[2] = 3
+   * RankTwoTensor A;
+   * A.fillFromScalarVariable(v);
+   * // A = [ 1 3 0
+   * //       3 2 0
+   * //       0 0 0 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   *
+   * When `input.size() == 6`, the second order tensor is filled symmetrically using the Voigt
+   * notation:
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+   * // Suppose v[0] = 1
+   * //         v[1] = 2
+   * //         v[2] = 3
+   * //         v[3] = 4
+   * //         v[4] = 5
+   * //         v[5] = 6
+   * RankTwoTensor A;
+   * A.fillFromScalarVariable(v);
+   * // A = [ 1 6 5
+   * //       6 2 4
+   * //       5 4 3 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
+  void fillFromScalarVariable(const VariableValue & scalar_variable);
```

Comment:
Sidebar: can we combine `fillFromInputVector` and `fillFromScalarVariable` ? (I thought I had already done that). If we make the argument type a template type and take advantage of _duck typing_. I.e. the passed in type must have `.size()` and `operator[]`. I thing Both `std::vector<T>` and `VariableValue` do. This would also enable use of `std::initializer_list` and `std::array` for example...